op-program/host/kvstore/disk.go

```diff
@@ -0,0 +1,67 @@
+package kvstore
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// read/write mode for user/group/other, not executable.
+const diskPermission = 0666
+
+// DiskKV is a disk-backed key-value store, every key-value pair is a hex-encoded .txt file, with the value as content.
+// DiskKV is safe for concurrent use; Puts may conflict, but write the exact same data anyway.
+type DiskKV struct {
+	path string
+}
+
+// NewDiskKV creates a DiskKV that puts/gets pre-images as files in the given directory path.
+// The path must exist, or subsequent Put/Get calls will error when it does not.
+func NewDiskKV(path string) *DiskKV {
+	return &DiskKV{path: path}
+}
+
+func (d *DiskKV) pathKey(k common.Hash) string {
+	return path.Join(d.path, k.String()+".txt")
+}
+
+func (d *DiskKV) Put(k common.Hash, v []byte) error {
+	f, err := os.OpenFile(d.pathKey(k), os.O_WRONLY|os.O_CREATE|os.O_EXCL|os.O_TRUNC, diskPermission)
+	if err != nil {
+		if errors.Is(err, os.ErrExist) {
```

Comment:
Even with this it's possible for concurrent calls to `Put` and `Get` to overlap such that the `Get` will return that the file exists but is empty (or get partial content).

To confirm I shoved a `time.Sleep(15 * time.Second)` before the `f.Write` call just below and ran:
```
func TestName(t *testing.T) {
	tmp := t.TempDir() // automatically removed by testing cleanup
	kv := NewDiskKV(tmp)

	hash := common.Hash{0xaa}
	data := []byte{1, 2, 3, 4, 5}
	go func() {
		err := kv.Put(hash, data)
		require.NoError(t, err)
	}()
	time.Sleep(3 * time.Second)
	res, err := kv.Get(hash)
	require.NoError(t, err)
	require.Equal(t, data, res)
}
```
the Get call winds up returning no error and empty data.

Reply:
Added a rw lock and updated docs to cover usage safety.